Add bary/heliocentric radial velocity correction

eteq opened the pull request:
This PR implements functions for computing heliocentric velocities.  The main practically useful bit is the `helio_corr` function.  That computes the correction required to convert an observed radial velocity to a heliocentric velocity.  This is a bread-and-butter calculation for basically all astronomical spectroscopic analysis.  The neat thing is that the implementation here is pretty trivial now that the SS ephemerides stuff is in, and can even use JPL ephemerides if needed.

Now the question is: is this how we want to deal with velocities?  Instead of doing these as piecemeal functions, this could be done as an actual transformation framework.  That needs to be a separate discussion (I think there's already an issue about it?), but either way this code is fundamental.  So I'm creating this PR now so that the code is here, and if we can get this figured out by 2.0 we put it in that framework, otherwise we merge this for 2.0 to make sure *something* is in there to do this fundamental activity.

One bonus bit to add *if* we go this route
- [x] add Solar System barycentric option - that's actually *simpler* than this, involving just skipping the bary->sun conversion step (essentially a 1-line change)

cc @adrn @mhvk @astrofrog

eteq commented:
Hmm, a curious failure mode to `matrix_product` is revealed in the tests.  @mhvk, do you have any idea what's going on in https://travis-ci.org/astropy/astropy/jobs/196493982 ?

mhvk commented:
Haven't looked in much detail yet, but it looks like the inputs to `matrix_product` are not, in fact, matrices!

mhvk commented:
Just to add back links: issue asking for barycentric correction: #3544; @StuartLittlefair's sample code: https://gist.github.com/StuartLittlefair/5aaf476c5d7b52d20aa9544cfaa936a1

(I'll try to look at the actual code here soon...)

mhvk commented on astropy/coordinates/velocities.py:
```diff
@@ -0,0 +1,74 @@
+    gtarg = target.transform_to(GCRS(obstime=t, obsgeoloc=gcrs_p))
+    targxyz = gtarg.represent_as(UnitSphericalRepresentation).to_cartesian().xyz
+
+    res = matrix_product(vsuntarg_cartrepr.xyz, targxyz)
```
OK, what you really want here is the regular dot product, i.e., `np.inner`. Annoyingly, that function is not quantity-proof.

*BUT* we just changed representations to do this for us, so just do
```
gtarg /= gtarg.norm()
res = gtarg.dot(vsuntarg_cartrepr)
```

StuartLittlefair replied:
@mhvk is almost right. Since you've already normalised ```gtarg``` through the conversion to a ```UnitSphericalRepresentation``` and back, you want:

```python
targcart = gtarg.represent_as(UnitSphericalRepresentation).to_cartesian()
res = targcart.dot(vsuntarg_cartrepr).to(KPS)
```

mhvk replied:
In my suggestion I wrote `gtarg` on purpose, i.e., I assumed the whole line `targxyz` would be removed (as it is unnecessarily slow compared to just dividing by the norm).

eteq replied:
I went ahead and profiled it, and, quite surprisngly, the `UnitSphericalRepresentation` way is *faster*:
```
>>> %%timeit 
>>> targcart = gtarg.represent_as(UnitSphericalRepresentation).to_cartesian()
>>>res = targcart.dot(vcorr_cart)
497 µs ± 14.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
and
```
>>> %%timeit 
>>> gtargd = gtarg.data
>>> gtargd /= gtargd.norm()
>>> res2 = gtargd.dot(vcorr_cart)
891 µs ± 28.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
That surprises me, but there it is... so will go that route.

mhvk replied:
OK, real data beats everything! I also think the route is also clearer than the original - it still avoids the expensive `.xyz`. Note that you can remove the `.to_cartesian()` -- `.dot` will do that for you.

But of course need to understand: `represent_as(USph)` is fast - just a new class initialization (this can become faster still...). Then, `to_cartesian` is slightly faster than it would have been before, since no multiplication with a distance (of 1.). While for my original suggestion there is the division by the distance (quite fast, but needless), a similar reinitialization, and then a slightly more expensive conversion to cartesian. My timings are not as different, but still the route via unitspherical is clearly better:
```
g = SphericalRepresentation(np.arange(0,100)*u.deg, np.arange(-45., 55.)*u.deg, np.arange(100,200)*u.kpc)
v = CartesianRepresentation(5., 5., 5., unit=u.km/u.s)
%timeit (g/g.norm()).dot(v)
# 1000 loops, best of 3: 1.69 ms per loop
%timeit g.represent_as(UnitSphericalRepresentation).dot(v)
# 1000 loops, best of 3: 1.49 ms per loop
%timeit g.represent_as(UnitSphericalRepresentation).to_cartesian().dot(v)
# 1000 loops, best of 3: 1.49 ms per loop
```

StuartLittlefair commented:
Hi @eteq,

I think it could be a good idea to implement this code now and worry about integration with a framework later. I'll add some inline style comments to the code, but I have a larger concern about precision.

If you read [this paper](http://iopscience.iop.org/article/10.1086/678541) - and in particular pp 842, you'll see that the approach we're taking here of adding and subtracting velocities, cannot be more accurate than a few m/s. Should we be aiming for better than this? Without cm/s precision, the exoplanet folk won't be interested in this, whereas the time corrections are good enough for all but pulsar astronomers. Given the exoplanet community is possibly the largest user base for this code, I'd argue we should be implementing a more precise algorithm.

StuartLittlefair commented on astropy/coordinates/velocities.py:
```diff
@@ -0,0 +1,74 @@
+    gcrs_p, gcrs_v = loc.get_gcrs_posvel(t)
+
+    vsuntargxyz = vsunearth.xyz + gcrs_v.xyz
+    return CartesianRepresentation(vsuntargxyz.to(KPS))
```
This isn't needed any more, you should be able to do:

```python
return vsunearth + gcrs_v
```

unless you really want it in km/s at this point

StuartLittlefair commented on astropy/coordinates/velocities.py:
```diff
@@ -0,0 +1,74 @@
+    vsuntarg_cartrepr = helio_vector(t, loc)
+    gcrs_p, _ = loc.get_gcrs_posvel(t)
+
+    gtarg = target.transform_to(GCRS(obstime=t, obsgeoloc=gcrs_p))
```
For some reason I find this confusing, but shouldn't you be using the ```ICRS``` co-ordinate of the target here, and finding the dot product with the velocity vector? It will only make a ~1 m/s difference but it will be important if we shoot for a more precise algorithm

eteq replied:
Hmm... I don't think that's right.  It's sort of in the definition of the correction, but the way I think about barycentric correction is: "the velocity vector of the earth relative to the barycenter, projected along the line-of-sight from the observer to the target".  If I used the ICRS coordinate instead of the GCRS coordinate, that would be the line-of-sight from the *barycenter* to the target. For anything beyond the solar system this is a tiny effect... but it is still what I would say is the correction.

Should probably be explicit about this in the docs, though.

mhvk replied:
I think @eteq is right: consider a not too-distance star that is exactly at the ecliptic pole, and which has a proper motion. As the earth orbits the sun, this proper motion would induce a radial-velocity modulation, because as seen from the geocenter, the position of the star oscillates around the ecliptic pole. But in ICRS, the star doesn't move.

StuartLittlefair commented on astropy/coordinates/velocities.py:
```diff
@@ -0,0 +1,74 @@
+    return CartesianRepresentation(vsuntargxyz.to(KPS))
+
+
+def helio_corr(t, loc, target):
```
In my gist I used the signature ```(Time, SkyCoord, EarthLocation=None)```, to be consistent with how we implemented time corrections. Again, for consistency, if the ```EarthLocation``` is ```None```, we try and get the location from the ```Time``` object.

eteq replied:
I'm actually really against any location information from the `Time` object.  The whole point of having `EarthLocation` existing is to get around the awkwardness there, and its more natural anyway for something in `coordinates`.  But the rest sounds fine!

mhvk replied:
We had a long discussion for the barycentric time corrections about where these things should live. For velocity corrections, to me `SkyCoord` is more logical, in which the signature would become `(SkyCoord, Time, Location)` - with defaults of `sc.obstime` and `sc.location` (and, if a time is passed in, a secondary default of `time.location`).

That said, I think it would be fine to have a short-cut on `Time` (and similarly a short-cut for time corrections on `SkyCoord`).

mhvk replied:
If we do add it to `Time` as well, I would stick to @StuartLittlefair's signature for consistency with `light_travel_time` - it seems logical to look for defaults first on `self`.

mhvk commented:
On the implementation: I would suggest initially to follow the precedent of the barycentric time corrections, and make the public API be a method rather than a stand-alone version. The method is probably most logical on `SkyCoord` (and one may want the ability to calculate the time difference there too, even though it breaks the "only one way" rule). If so, the most logical signature might be
```
def radial_velocity_correction(self, obstime=None, kind='barycentric', location=None, ephemeris=None)
```
where `obstime` would by default get taken from the coordinate, as would `location` (from `obsgeoloc`, with `obstime.location` as a secondary default). Not completely sure whether it should be `location` or `obsgeoloc` -- I do not like the latter for its abbreviation, but it is what is stored in some `SkyCoord`.

mhvk commented:
Longer term, I guess we'd want `SkyCoord` to be able to have both proper motion and radial velocity. It may get slightly tricky to deal with situations where some of those components (and/or distance) are missing, but we have to face this one way or the other...

mhvk commented:
@StuartLittlefair - Looking at the reference you cited, it would seem what we need to provide is just the velocity correction, but we have to make clear in the documentation that one applies this as
```
v_corrected = c*[(1+v_measured/c) * (1 + v_correction/c) - 1.]
```
(and we'll need to implement that ourselves later...)

StuartLittlefair commented:
Yes, that's true. 

However, I think we are still missing important (i.e m/s) physics. There's a [python interface](https://github.com/tronsgaard/barycorr) to the online calculator I linked to earlier. If you install that you can see that neither my gist, nor @eteq 's code agrees with the results of BARYCORR to a few m/s.

[This gist](https://gist.github.com/StuartLittlefair/640d1f8a0429cbfbb3de91d3dc52e29c) shows an example...

mhvk commented:
@eteq - changed the milestone to 3.0 for now -- I think we have enough on our plate without this (and if you get the velocities right, this comes for free with the proper coordinate transform...).

eteq commented:
Note that, while in *principal* this functionality is now supported (with a lot more) in #6226 and the PR's that it are derived from, the finite difference calculation in that PR has some numerical stability issues for *some* relevant use cases.  So I think the best thing to do is include this PR (updated with the comments, of course), with a note in the docs that it may at some point be adjusted to use the full velocity framework if the numerical stability problems can be solved.

Will update this PR ASAP with that plan in mind.

StuartLittlefair commented:
Great @eteq - perhaps add a note in the docs that this PR doesn't yet provide m/s precision and link to the barycorr Python module? Otherwise there's a risk we'll get lynched by some exoplanet people...

mhvk commented:
I'm still rather uncomfortable with the API here, with a new module. Since this is rather simple, I'd much rather have a method on `SkyCoord` (or possibly `Time`). 

More generally, I think with the velocities we have more than enough on our plate, and I prefer to think about this better -- another option is to simply document this as part of `solar_system`.

eteq commented:
I'm about to try to re-work this with the implementation comments above in mind, but just a couple thoughts on the bigger picture from @StuartLittlefair:
* I think reaching the cm/s precision will take a lot more care than we have done thus far.  In particular, most of our ephemeides aren't better than ~1 m/s right now.  I think this would be a great project to sic someone on, though... once we have the *start* of the framework in hand.
* Agreed that this means it should come with a clear warning of its precision level.  But actually there are plenty *lot* of use cases for this that are not exoplanet folks - pretty much anyone who wants to do Galactic or extragalactic work with spectra has to do this, and km/s level precision is usually more than sufficient for those cases. So hopefully we can get them something now, and build up to the higher precision case.

eteq commented:
@mhvk - I'm receptive to the idea of it being in a method, as you suggest... the main concern, though, is that it might lead to a lot more confusion with the velocities work @adrn and I are trying to finish.  A function, meanwhile, is much easier to cast as sort of a "temporary" workaround or an alternate approach (even if we support it indefinitely by having it eventually be based on the properly-done coordinates stuff).

eteq commented:
Ok, I pushed up a new version, based in part on @mhvk and @stuartLittleFair's feedback.

Some significant differences:
* the public API is now as a method on `SkyCoord` as Marten suggested
* `obstime` and `location` are *not* valid arguments.  Instead I'm *requiring* obstime and location to be in the SkyCoord.  I think that's not unreasonable/burdensome and it seems to follow the only-one-way-to-do-it rule better.
* `ephemeris` is not a keyword, rather there's a note that you can mess with the `solar_system_ephemeris` if you want to do so.

Other than that I think it adheres pretty will to the last version + feedback.

bsipocz commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1225,6 +1225,61 @@ def from_pixel(cls, xp, yp, wcs, origin=0, mode='all'):
+                             "be 'barycentric' or 'heliocentric', but got "
+                             "'{}'".format(kind))
+
+        return rvcorr(self.obstime, self.location, self, bary, ephemeris)
```
ephemeris is left in here but isn't defined any more

mhvk commented:
Looking at this again, a larger point: almost always one wants `barycentric` velocity corrections; I think these might as well be provided too, with a switch like for `Time.light_travel_time`.

p.s. It is these kinds of considerations why I'm a bit hesitant to move ahead before the API is thought out well -- I think we've found by now how some API decisions, often made in pre-release haste, come back to haunt us.

StuartLittlefair commented on astropy/coordinates/tests/test_velocity_corrs.py:
```diff
@@ -0,0 +1,199 @@
+    vhs_iraf = vhs_iraf*u.km/u.s
+
+    vhs_astropy = radial_velocity_correction(IRAF_time, loc, coos)
+    assert_quantity_allclose(vhs_astropy, vhs_iraf, atol=150*u.m/u.s)
```
150 m/s is quite a large discrepancy, even for non-planetary work. For my own work on binaries, we routinely measure semi-amplitudes and systemic velocities to 100 m/s acuracy, so if the discrepancies are on this level it's an issue to many, and not compatible with the 1-10 m/s quoted in the docstring.

Is this our problem, or IRAFs?

eteq replied:
I'd say IRAF's.  I compared the result of the barycentric correction to BARYCORR and it is right to 1-2 m/s (which I think you noted above).

My memory is a bit hazy now, but as I recall the IRAF algorithm is a bit ambiguous about how it defines some of its definitions relative to the modern way (or at least *I* didn't understand it). I remember thinking at the time, at least, that this didn't surprise me because of of imperfect IRAF-something. But at the *time* I didn't have a better one.  Of course BARYCORR counts as that now.

eteq replied:
( Oh, and inspired by this I will add a BARYCORR-driven test)

StuartLittlefair commented:
I do wonder if @mhvk has a point? Given the awkwardness of deciding where this method should sit, the fact that we may deprecate it at some point, and that the part of the code that does the work is only a few lines, I wonder whether it would be better to add this approach to the documentation instead?

mhvk commented:
I realize I wrote an in-line comment before seeing what had changed.

On the API: I appreciate the strictness, but will argue for allowing one to pass in time, location, and possibly ephemeris: generally, people observe a given source many different times. So, they have a scalar `SkyCoord`, likely ICRS, an array of times, and probably a single location. Given that neither `obstime` nor `location` are natural for ICRS, why force them to pass this in the initializer?

In fact, it is a bit worse: while currently one can actually initialize a scalar ICRS `SkyCoord` with an array of `obstime`, that is actually a bug, since any conversion to a frame that does require `obstime` will fail. This is because the current code insists that we would not broadcast the coordinates proper, and cannot have have attributes with a shape inconsistent with the coordinates (indeed, initialization of an `FK4` coordinate with an array of `obstime` fails for this reason). [Of course, we can revisit this; in some sense, it is this way to mimic `Time`, where a single time with, say, multiple locations really doesn't make much sense.]

Anyway, I'd still argue for (though not strongly for `ephemeris`):
```
def velocity_correction(self, kind='barycentric', time=None, location=None, ephemeris=None)
```
I also think we should add a `light_travel_time` method which just calls the `Time` one (or, perhaps more logically given the many in-method imports in the current implementation, move the `Time` method to here and let the one on time call the moved one). While the only-one-way is fine, I think for this case the choice is so unobvious that I think it is quite OK for `Time` to gain a `velocity_correction` method as well, to complement `light_travel_time`.

p.s. Actually not quite sure about best name either: maybe `observer_velocity`, `observer_rv`, `rv_correction`?

mhvk commented:
On the path forward: I think we should only merge once we have precisions close to that of `epv00`, i.e.,  better than `1 cm/s` for barycentric (and probably heliocentric too). This probably means not for 2.0...

StuartLittlefair commented on astropy/coordinates/velocity_correction_funcs.py:
```diff
@@ -0,0 +1,102 @@
+        vcorr_cart = helio_vector(t, loc)
+
+    gcrs_p, _ = loc.get_gcrs_posvel(t)
+
```
Another slight issue with this being a method on SkyCoord and using SkyCoord's location and obstime is that if one does set a SkyCoord with multiple obstimes, and directions but a single location, this transformation will fail

mhvk replied:
That actually turns out not to be true! (We changed `get_itrs` to handle multiple obstime):
```
EarthLocation.of_site('keck').get_gcrs_posvel(Time(50000., np.arange(10), format='mjd'))
```
works.

StuartLittlefair replied:
Sorry - wrong line! It's this one that will fail:

```gtarg = target.transform_to(GCRS(obstime=t, obsgeoloc=gcrs_p))```

E.g

```python
lapalma = coord.EarthLocation.of_site('lapalma')
times = Time(np.random.uniform(2457784, 2457785, 1000), format='jd')
ras, decs, _ = randomly_sample_sphere(1000)
Targets = SkyCoord(ras, dec, obstime=times, location=lapalma)
Targets.radial_velocity_correction()

File "/Users/sl/anaconda3/envs/astropy_testing/lib/python3.6/site-packages/astropy/coordinates/sky_coordinate.py", line 1281, in radial_velocity_correction
    return rvcorr(self.obstime, self.location, self, bary)
  File "/Users/sl/anaconda3/envs/astropy_testing/lib/python3.6/site-packages/astropy/coordinates/velocity_correction_funcs.py", line 100, in radial_velocity_correction
    gtarg = target.transform_to(GCRS(obstime=t, obsgeoloc=gcrs_p))
  File "/Users/sl/anaconda3/envs/astropy_testing/lib/python3.6/site-packages/astropy/coordinates/sky_coordinate.py", line 461, in transform_to
    new_coord = trans(self.frame, generic_frame)
  File "/Users/sl/anaconda3/envs/astropy_testing/lib/python3.6/site-packages/astropy/coordinates/transformations.py", line 935, in __call__
    curr_coord = t(curr_coord, curr_toframe)
  File "/Users/sl/anaconda3/envs/astropy_testing/lib/python3.6/site-packages/astropy/coordinates/transformations.py", line 738, in __call__
    res = self.func(fromcoord, toframe)
  File "/Users/sl/anaconda3/envs/astropy_testing/lib/python3.6/site-packages/astropy/coordinates/builtin_frames/icrs_cirs_transforms.py", line 129, in icrs_to_gcrs
    axis=-2)
ValueError: all the input arrays must have same number of dimensions
```

mhvk replied:
That would seem like a bug. The underlying problem is that `obsgeoloc` is defined but not `obsgeovel` -- before trying to fix the bug, should that in fact raise an error?

For here, the solution would seem to be to not throw away the velocity information, but rather use it as well.

Possibly, `EarthLocation` should just have a `get_gcrs` method that returns this directly?

eteq replied:
Good point on using the velocity information!

I could see the argument for a `get_gcrs`, *but* my concern is that it's confusing that it's subtly different from the `EarthLocation.get_itrs` - the latter gets the *position* of that `EarthLocation` as an `ITRS`, while the `get_gcrs` here is a GCRS frame with the *origin* at that `EarthLocation`.  In any case, I'd rather not get too caught up in that for this PR.

eteq commented:
@mhvk
> On the API: I appreciate the strictness, but will argue for allowing one to pass in time, location, and possibly ephemeris: generally, people observe a given source many different times. So, they have a scalar SkyCoord, likely ICRS, an array of times, and probably a single location. Given that neither obstime nor location are natural for ICRS, why force them to pass this in the initializer?

I don't understand why you say this - `obstime` and `location` seem very natural here *on a `SkyCoord`* (as opposed to a frame object).  After all, the first step in the core algorithm is to convert `target` to `GCRS`, which *requires* an `obstime`.  In fact, as a `SkyCoord` method, all that I "should" have had to do was ``target.gcrs`` for it to work.  This implementation I think is a bit more explicit and better for testing anyway, but it is effectively identical to what `SkyCoord` would do internally. I see this as the whole point of `SkyCoord` as opposed to using frame classes directly: `SkyCoord` is there to store attributes that might need even by a part of the current frame *because* you're going to translate it into another frame later.

> while currently one can actually initialize a scalar ICRS SkyCoord with an array of obstime, that is actually a bug

This had better not be true!  Quite a bit of code in the wild (including most of the `astroplan` affiliated package) assumes that it is possible for `obstime` to be an array and that it will broadcast. So that would be a major API-break to change now.  After all, if ``np.array(10) + np.arange(10)`` works, why shouldn't this?

I've added some tests in the code that I pushed up that runs the code in the array-obstime mode.  It does appear to work right now using *only* the `SkyCoord` interface, so I still favor that.  If we really don't want to use the `SkyCoord` frame, though, I'd say it *should* be a function and not a method...

eteq commented:
> On the path forward: I think we should only merge once we have precisions close to that of epv00, i.e., better than 1 cm/s for barycentric (and probably heliocentric too). This probably means not for 2.0...

I very much disagree with this.  I don't think we should target the state-of-the-art as our boundary for including something... If we did most of Astropy wouldn't exist.  I am fine with putting in a clear warning (as is already there), but I think a precision of ~1 m/s (which is what this has) is still very useful for many people.  I've seen lots of people with their own version of this code (including @stuartlittlefair !) which seems to me a clear sign that we need this in Astropy to prevent further unnecessary incompatible proliferation.

adrn commented:
For what it's worth, I do think this should be in v2.0 (with a warning about precision). I'll try to get to a full review by tomorrow.

mhvk commented:
@eteq - hmm, I was wrong in my statement, it is just the initializer that fails if you try to put in an array of obstime and a single coordinate for a frame that cares about obstime:
```
sc = SkyCoord(0.*u.deg, 0.*u.deg, obstime=Time(np.arange(50000, 50010), format='mjd'), frame='fk4')
ValueError: cannot broadcast a non-scalar to a scalar array
```
This is a somewhat separate issue (#5982).

mhvk commented:
On the precision: if we've got down to 1 m/s, yes, I think that is sufficient for now -- but let's agree that improvements in that from small changes count as bugs! I also think we do not have to warn about deprecation -- this method is going to stay as a convenience.

I'm *still* in favour of allowing `obstime`, `location`, and `ephemeris` as arguments! To try to convince you this is a useful convenience, and that any other route is not all that obvious, let me describe how as a not-even-naive user I would approach this (it is what I actually tried...):
```
# My observed source, set at the start of my reduction script
sc = SkyCoord.from_name('PSR J1012+5307')
# later in the program I want rv corrections at `t`, `loc`
# the location is probably put in by hand, the times from
# headers (but not coordinates -- wrong too often).
loc = EarthLocation.of_site('keck')
t = Time(50000., np.arange(10), format='mjd')
# So, let me just set those on sc...
sc.location = loc
# Works
sc.obstime = t
# Oops
# ValueError: cannot broadcast a non-scalar to a scalar array
```
(This is not going to get fixed by #5982, since one cannot change the shape of a `SkyCoord` after creation as it is meant to be immutable...)

So, I guess we need to wrap:
```
sc2 = SkyCoord(sc, obstime=t, location=loc)
sc2.rv_correction()
# Yay! (didn't install your PR, so suppose it worked)
# But let me just check what I actually used:
sc2
# <SkyCoord (ICRS): (ra, dec) in deg
#    ( 153.1393271,  53.117343)>
# Hmm, obstime and location are not shown.
sc2.shape
# ()
# Hmm, really no hint that there is a higher dimension here.
```

I think the above route is sufficiently unexpected that the docstring for `rv_correction` would need to describe it in some detail. And even then I rather dislike that we are working with "hidden" state of an object (indeed, I dislike this enough to raise #6283).

So why not just provide the option to the user? If one has to choose between the following, is there a real reason to prefer the first?
```
# eteq route:
sc2 = SkyCoord(sc, obstime=t, location=loc)
with coord.solar_system_ephemeris.set('jpl')
    rv += sc2.rv_correction()

# mhvk route
rv += sc.rv_correction('barycentric', obstime=t, location=loc, ephemeris='jpl')
```
(Note that to write the "eteq" route I actually had to check the documentation on how to set the solar system ephemeris, since the current docstring does not mention the ".set" part -- and I *wrote* half of that code!)

Now I realize you argued one might as well use a function, and that's true, and really my only argument against that is that it is less discoverable than a method (indeed, discoverability by tab completion argues for `radial_velocity_correction`). So, to me the above is just the right mix of functional and object-oriented programming... And, of course, there is the precedent of `Time.light_travel_time`...

adrn commented:
I haven't yet done a line-by-line review of this, but in reading the last few back and forth comments I come out with a slight preference for @mhvk's method with arguments idea. To me it seems unnatural to pass in an `obstime` and `location` to `SkyCoord`. For one, the coordinates of an object shouldn't care about the `location` (isn't that why we have the ICRS?), so it feels awkward. Conceptually, I like the idea that I can load / define the coordinates for my object, then later handle what to do once I "observe" it from a place and time. 

I'm also somewhat leery of naming this `rv_correction`. For consistency with the frame attributes I added in #6219, it should at least be `radial_velocity_correction`. Even then, I don't think of it as a correction...but don't have a better word for you.

eteq commented:
> On the precision: if we've got down to 1 m/s, yes, I think that is sufficient for now -- but let's agree that improvements in that from small changes count as bugs! I also think we do not have to warn about deprecation -- this method is going to stay as a convenience.

👍 

On the arguments or not question:

I can accept being outvoted here.  I see the point that it is a "convenience" method...  That creates some additional questions, though:

- [ ] If a `obstime`/`location` is given in the method and there is a *different* one on the `SkyCoord`, is that an error? I assume the answer is "no" based on the above use cases (if "yes", then the other two questions are moot points).
- [ ] What should be done if the `SkyCoord` has an `obstime` or `location` and its part of the current frame?  Do keyword arguments mean "override that before the transformation to GCRS", or "use that *as* the transformation to GCRS"?  Those are subtly but very importantly different for some frames (most practically, `AltAz`).  The former (re-assign the SkyCoord attributes a la @mhvk's example) seems better so I will try implementing that but am not sure if that is really the most natural.
- [ ] What should be done if the `SkyCoord` has an `obstime` or `location` and its *not* part of the current frame?  Again, the "before" or "after" question applies, although its somewhat less pressing in this case (for e.g. ICRS it doesn't really matter because the only frame in the transform chain with an obstime is `GCRS`).  "Before", again, seems most natural to me.

(Also, to be a bit stubborn... the fact that we need to ask these questions is precisely why I was in favor of not allowing this - I'm pretty sure different users will assume different things on this, but if its all in `SkyCoord` then we don't need to worry about that - the rules are already clear: it behaves "like a `SkyCoord`".)

mhvk commented:
@eteq - good point, of course...  For `Time.light_travel_time` we allow overriding the location, but then, it is not a big effect -- here, I think this is not the case. I think I prefer to raise in all cases at least initially, but maybe good to think through what the rule should be if we wouldn't. The only one that makes sense to me is to first transform the `SkyCoord` to ICRS with its internal properties, and the use `location` and `time` as given. I think, though, that this is too risky, so prefer just raising an exception.

mhvk commented:
Separately, right now the method simply calls a function. I think I'd prefer the method to *be* the implementation, but if we do not do that, I'd suggest to move the function to `solar_system` rather than create a new sub-module (`solar_system` includes velocities, so I think it is appropriate). Indeed, in that case I'd be in favour of also moving `light_travel_time` there.

@StuartLittlefair - what do you think about adding these to `solar_system`?

eteq commented:
@mhvk - 
> I think I prefer to raise in all cases at least initially, but maybe good to think through what the rule should be if we wouldn't

Hmm... that surprises me a bit because that would break exactly the use case you cited above except in the case when the `SkyCoord` is ICRS? (or maybe that's your point with the `ICRS` bit?) So then I come back to why do it at all if it raises errors for one of what I would have thought would be a solid use case?

But I see your point that this is the most "conservative" one.  It's also has the advantage of being the easiest to implement so I'll try to get that up ASAP.


> I'd suggest to move the function to solar_system rather than create a new sub-module

I don't think it belongs in `solar_system`, because most of the use cases for the barycentric correction that I know of  are *not* really solar system. My reason for not putting it in `SkyCoord` directly is entirely that `sky_coordinate.py` is alright full of a ton of stuff, and doing the function separately makes it easier to understand by others (which might be especially important for this use case...). We have the precedent of the `separation*` and `match_to_*` methods for this already...   But then again its also true that those have a separate public API.  

Will try putting it in `sky_coordinate.py` and see if it seems too awkward...

eteq commented:
Alright, the version I pushed up now has some changes based on the feedback above.

First, having done it now, @mhvk was absolutely right this is better as a stand-alone method.  Turns out that was more straightforward than the original module...

Second, while they're now in there, I'm still not sold in the `obstime` and `location` arguments even having implemented them (although it's all in one commit, so easy to back out!).  With the "raises errors" version in place as it currently stands, it seems to me like this is just the same version *without* the arguments, but with a bunch of error checking to make sure no one puts in inconsistent arguments.  It seems just as natural, look at @mhvk's examples, to just say the user should do this:
```
vcorr = SkyCoord(sc, obstime=t, location=loc).radial_velocity_correction()
```

And while I agree with the point in https://github.com/astropy/astropy/pull/5752#issuecomment-309268570 that it's a bit odd that this is using "hidden state", hopefully a solution to #6238 will clear that up.

Part of the reason I'm pushing on this is that it seems like this is making a choice now that might lead to more confusion we can't back out of later, while sticking with only the `SkyCoord` attributes allows us to later add these arguments if it's clear there's call for them?  (the other way is much harder because it requires removing something.)  

All that whining said: I think this is ready to go in if I still haven't convinced anyone, hopefully this is good to go (and if I have... easy enough to drop the last commit :wink:)

StuartLittlefair commented:
@eteq to me it's simply a question of convenience so I'd be happy for you to drop the arguments and document the approach in your comment provided it isn't much slower for large numbers of targets and/or times. I suspect the initialisation of a new skycoord might be a performance hit?

eteq commented:
@StuartLittlefair - for "large numbers" you mean lots of `SkyCoord` *objects* or array-SkyCoords?  
 The latter it should be negligable, but the former would definitely be noticeable... although I guess it's probably at the worst a factor of 1.5x because the GCRS transform is required regardless.

eteq commented:
That is to say: if performance of this type is a significant concern then that is an argument for the "argument: version.

(Although one of the things I think we want to focus on for v3.0 is a big push on performance that *hopefully* would help matters a lot in that department)

StuartLittlefair commented:
For my own use case I would be attaching a large number of times to a skycoord that has one coordinate

eteq commented:
Ohh, yeah, in that case it should be negligible because you can use an `obstime` array with a single `SkyCoord` (assuming "large number" is big enough to swamp a few SkyCoord initializations)

StuartLittlefair commented:
In which case I've little objection to the approach you prefer, apart from the inconsistency with ```light_travel_time``` which bugs me, but I can live with it.

mhvk commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1225,6 +1225,92 @@ def from_pixel(cls, xp, yp, wcs, origin=0, mode='all'):
+
+        t = self.obstime
+        if (obstime is not None and self.obstime is not None and
+           obstime != self.obstime):
```
This will fail if one or both are array-valued (returns an array of bool, or even errors if shapes are inconsistent); I'd just remove this -- up to the user to not pass in something if `self.obstime` exists.

mhvk replied:
The structure of the if-else seems odd; would do
```
if obstime is None:
    obstime = self.obstime
    if obstime is None:
        error
elif self.obstime is not None:
    error
```
and work with `obstime` below.

eteq replied:
I see your point here, yeah, it actually does make sense to *not* do this check after-all.

adrn replied:
Nitpick: align `obstime` in if statement so the indentation is pep8 compliant

mhvk commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1225,6 +1225,92 @@ def from_pixel(cls, xp, yp, wcs, origin=0, mode='all'):
+                             'radial_velocity_correction, either as a SkyCoord'
+                             ' frame attribute or in the method call.')
+
+        loc = self.location
```
Similar (do we want to allow `location` to be in the `time`?)

eteq replied:
Ugh, forgot that `time` can have a location :( I lean towards failing *no matter what* if `time` has a location because I think it's a lot more confusing to keep all these together.  Actually technically it might be more correct to set the location on the `Time` before doing other things... but I *think* that doesn't matter anywhere in this calculation right now.

mhvk replied:
Maybe do location before time: if not given, try `self.location`, if that is `None`, try `obstime.location`, if that is `None` fail. I would ignore any location in `self.time`, though (i.e., neither ever use it, nor fail if it has one).

mhvk commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1225,6 +1225,92 @@ def from_pixel(cls, xp, yp, wcs, origin=0, mode='all'):
+        The default is for this method to use the builtin ephemerides for
+        computing the sun and earth location.  This can be changed by setting the
+        `~astropy.coordinates.solar_system`_ variable, either directly or via
+        ``with`` statement.
```
If you don't want to use the `ephemeris` argument, can you expand this note to tell *explicitly* how to use `solar_system`, so that someone doesn't have to look this up?

eteq replied:
Good point... will mull a bit, bit will do one or the other.

adrn replied:
I don't have a strong opinion about an `ephemeris` argument (I abstain!) but I do agree that if you don't add it, you should be more verbose here and possibly put a code example about how to use the `with` statement to change it.

adrn commented:
Maybe I misunderstood something earlier, but I thought the debate was:
* Allow `obstime` / `location` via `SkyCoord` initializer, __OR__
* Allow `obstime` / `location` via arguments to `.radial_velocity_correction`

Not:
* Allow `obstime` / `location` via `SkyCoord` initializer, __OR__
* Allow `obstime` / `location` via `SkyCoord` initializer and via arguments to `.radial_velocity_correction`
(what your last few comments imply)

?

eteq commented:
@adrn - actually it's 
* Allow `obstime` / `location` via `SkyCoord` initializer, __AND__
* Allow `obstime` / `location` via arguments to `.radial_velocity_correction`
__OR__
* Allow `obstime` / `location` via `SkyCoord` initializer, __ONLY__

mhvk commented:
@adrn - @eteq said it right, one has the option to provide arguments or not provide any, just like for `Time.light_travel_time` -- best of both worlds. Though I think we all agreed that we would exclude getting some information from each.

adrn commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1225,6 +1225,92 @@ def from_pixel(cls, xp, yp, wcs, origin=0, mode='all'):
+                                   location=None):
+        """
+        Compute the correction required to convert a radial velocity at a given
+        time and place to a barycentric or heliocentric velocity.
```
Time and place *on Earth's surface*?

adrn commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1225,6 +1225,92 @@ def from_pixel(cls, xp, yp, wcs, origin=0, mode='all'):
+
+        Parameters
+        ----------
+        kind : str
```
I hate string arguments (just needed to get that out). 

But seriously, is there anything we can do to not have to ask people to type and spell `'barycentric'`? I don't have any other immediate ideas...so probably the answer is no!

eteq replied:
No.  You're the one who always wants long names! 😉 

Seriously, though, I agree but couldn't come up with anything better either, other than `bary=True`/`bary=False` which I dislike even more

adrn replied:
Did someone say "long names"?

`radial_velocity_correction_barycentric` / `radial_velocity_correction_heliocentric`

adrn replied:
`radial_velocity_barycentric`/ `radial_velocity_heliocentric`

adrn commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1225,6 +1225,92 @@ def from_pixel(cls, xp, yp, wcs, origin=0, mode='all'):
+            'heliocentric'.
+        obstime : `~astropy.time.Time` or None, optional
+            The time at which to compute the correction.  If `None`, the
+            ``obstime`` frame attribute on the `SkyCoord` will be used. (Raises
```
Should the parenthetical bit be moved to under a 
```
Raises
------
```
section?

eteq replied:
yes! good point

adrn commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1225,6 +1225,92 @@ def from_pixel(cls, xp, yp, wcs, origin=0, mode='all'):
+
+        Returns
+        -------
+        vcorr : astropy.units.Quantity with velocity units
```
Backticks? Also: this makes me remember that in Gala, when something accepts or returns a quantity with a particular physical type, I document that as (speed is the unfortunate name of the physical type for velocity):
```
vcorr : astropy.units.Quantity [speed]
```
Do we have a standard for this in Astropy?

eteq replied:
We do not have a standard for it... does that render `speed` in any special way, or just as-is?

adrn replied:
Hmm, makes me think that maybe we should. No, it doesn't natively. But is it possible to add an extension to numpydoc to check for such a thing?

adrn commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1225,6 +1225,92 @@ def from_pixel(cls, xp, yp, wcs, origin=0, mode='all'):
+        location : `~astropy.coordinates.EarthLocation` or None, optional
+            The observer location at which to compute the correction.  If
+            `None`, the  ``location`` frame attribute on the `SkyCoord` will be
+            used. (Raises an exception if the `SkyCoord` and this argument are
```
Same comment as above - should the parenthetical clause be under a `Raises` section?

eteq replied:
ditto

adrn commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1225,6 +1225,92 @@ def from_pixel(cls, xp, yp, wcs, origin=0, mode='all'):
+        Notes
+        -----
+        The algorithm here is sufficient to perform corrections at the ~1 to
+        10 m/s level, but has not been validated at higher prevision.  Future
```
Typo: prevision -> precision. Although I personally sometimes get confused by "higher precision" because it's really a lower error. Maybe "better precision"?

adrn commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1225,6 +1225,92 @@ def from_pixel(cls, xp, yp, wcs, origin=0, mode='all'):
+        t = self.obstime
+        if (obstime is not None and self.obstime is not None and
+           obstime != self.obstime):
+            raise TypeError('cannot compute radial velocity correction if '
```
"Cannot"

adrn commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1225,6 +1225,92 @@ def from_pixel(cls, xp, yp, wcs, origin=0, mode='all'):
+        elif self.obstime is None:
+            t = obstime
+        if t is None:
+            raise ValueError('Must provide an `obstime` to '
```
Nitpick: backticks around obstime here, but not in exception above.

adrn commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1225,6 +1225,92 @@ def from_pixel(cls, xp, yp, wcs, origin=0, mode='all'):
+
+        loc = self.location
+        if (location is not None and self.location is not None and
+           location != self.location):
```
Align `location`'s

adrn commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1225,6 +1225,92 @@ def from_pixel(cls, xp, yp, wcs, origin=0, mode='all'):
+        loc = self.location
+        if (location is not None and self.location is not None and
+           location != self.location):
+            raise TypeError('cannot compute radial velocity correction if '
```
Cannot

adrn commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1225,6 +1225,92 @@ def from_pixel(cls, xp, yp, wcs, origin=0, mode='all'):
+                             ' frame attribute or in the method call.')
+
+        if kind == 'barycentric':
+            vorigintoearth = get_body_barycentric_posvel('earth', t)[1]
```
I would use underscores for any of these velocity variables `v_origin_to_earth` and `v_sun` etc. But maybe that's a style thing...

adrn commented on astropy/coordinates/sky_coordinate.py:
```diff
@@ -1225,6 +1225,92 @@ def from_pixel(cls, xp, yp, wcs, origin=0, mode='all'):
+        elif self.location is None:
+            loc = location
+        if loc is None:
+            raise ValueError('Must provide a `location` to '
```
Backticks here around location and not above

adrn commented on astropy/coordinates/tests/test_velocity_corrs.py:
```diff
@@ -0,0 +1,296 @@
+    targets = SkyCoord(_get_test_input_radecs(), obstime=test_input_time,
+                       location=test_input_loc)
+    vhs_astropy = targets.radial_velocity_correction('heliocentric')
+    # vhs_astropy = targets.radial_velocity_correction(test_input_time,
```
Commented code

eteq replied:
good catch... was meant to be there to test the argument handling, but I forgot to uncomment!

eteq commented:
With the commits I just pushed up I think I addressed all the comments.  I did not add an `ephemeris` keyword, but did add an example of how to use the existing machinery to use an alternate ephemeris.

I think this is very close now (assuming tests pass), but two remaining unresolved items:

- [ ] In the current version I am disallowing *any* `time.location`, on the assumption that that's just too many layers of indirection/ambiguity.  I *could* do the chain @mhvk suggested (location arg, if not: SkyCoord.location, if not: obstime.location, if not: SkyCoord.obstime.location).  That seems likely to cause a lot of unforeseen oddities, though... 

- [ ] In the latest commit, `obstime`/`location` *cannot* be passed in as arguments if they're on the `SkyCoord`.  But this means that if a SkyCoord is in a frame that *requires* `obstime` or `location` (most important of these is `AltAz` which needs both), there' s no way to change `obstime` other than changing it on the SkyCoord).  Are we all ok with that?  (I am, because I think it demonstrates the original reason for not having it on the method... but I just wanted to make sure we all understood the challenge here)

I think @adrn and @mhvk might want to respond to those, so won't merge now.  But this is very close so a clear candidate for backporting later this week, and hence I won't push the milestone.

StuartLittlefair commented:
On disallowing ```time.location```: my problem with that is that it is totally inconsistent with how we have implemented ```light_travel_time```. Indeed the docs instruct the user to put the location onto the ```Time``` object before calculating ```light_travel_time``` corrections. I can understand you wanting to enforce some rigour here, but I think the user is entitled to feel frustrated if their next step is to calculate RV corrections and they have make a new ```Time``` object with the location removed, and add it to their ```SkyCoord```. 

Other than elegance, is there any real issue with having a clear hierarchy, e.g ```SkyCoord.location > Time.location > location```, as long as it is clearly documented?

On the second restriction, I can only think of very contrived use cases where you would want to use a different ```obstime``` or ```location``` - at least for the ```AltAz``` case. Again, my issue here is one of convenience to the user.

mhvk commented:
@StuartLittlefair's point I think is valid: typical use in my "ICRS coordinate case" would be to calculate both barycentric corrections for the same `Time` -- I think we might as well support the `location` inside `Time`, although I think we're OK with insisting that `SkyCoord` does not have an `obstime`, and ignoring any `location` inside `SkyCoord.obstime` (since it should, if used, be already copied to `SkyCoord.location`).

eteq commented:
OK, I updated this so that it will use `obstime.location` if it's given and neither of the other are present.  Is this now what you were thinking, @StuartLittlefair?  I was not entirely clear if you wanted to allow there to be multiple `obstime`'s and just pick the one higher in "precedence", or if your `>` symbols meant a recommendation?

I'm agains the idea of allowing any combination (e.g. `obstime.location` and `SkyCoord.location` are both not None), because there's just no way to tell what the user intended - did they *mean* to override, or just not realize that there's a `location` on that time?  (similar for `SkyCoord`).  It's easy enough to get such objects from other contexts without specifically attaching the location, so I think it's better to be conservative,  After all, "In the face of ambiguity, refuse the temptation to guess." ™️...

eteq commented:
@mhvk - good point re: docs, but my plan was to include this in the section that is to be added #6219 - so that would be easiest post-merge, and possibly in between the beta and final release.  We should create an issue accounting for that once this is merged and make sure its milestoned 2.0.

adrn commented:
I'm fine with adding this to the (coming) velocity documentation, but yes, an issue milestoned 2.0 is required!

mhvk commented:
Same, fine to add later, as long as there is a 2.0 issue for it.

eteq commented:
I added a note about adding the documentation for this in #6260.  Assuming the tests pass, I think this is now ready!

eteq commented:
I'm going to go ahead and merge this, but @StuartLittlefair, if I was misunderstanding your intent above, feel free to raise it as a separate issue and/or a PR of your own.  If we can get to that by next Monday we can still pull it into 2.0 as that would be a "bug" in my interpretation of your comment :wink:

StuartLittlefair commented:
Apologies. Thought I'd replied - looks fine, merge away!

eteq commented:
I'm going to go ahead and merge this, but @StuartLittlefair, if I was misunderstanding your intent above, feel free to raise it as a separate issue and/or a PR of your own.  If we can get to that by next Monday we can still pull it into 2.0 as that would be a "bug" in my interpretation of your comment :wink:

eteq commented:
Whoops! Forgot the changelog... I think I'll just backport this to v2.0.0 now then and fix it directly in master to avoid confusion...

eteq commented:
(commits listed above are for the changelog in master, the backport of this PR to v2.0.x, and the backport of the changelog commit, respectively)
